File: yujin/셔틀버스.py
def solution(n, t, m, timetable):
    avaliable = []
    times = []
    start_time = 9 * 60
    
    for time in timetable:
        [hours, minutes] = time.split(":")
        times.append(60 * int(hours) + int(minutes))
    
    times.sort(reverse=True)
        
    for count in range(n):
        avaliable.append([])
        while len(times) > 0 and len(avaliable[count]) < m and start_time + count * t >= times[-1]:
            avaliable[count].append(times.pop())
    
    print(avaliable)
    
    if len(avaliable[-1]) < m:
        return formatted(start_time + (n - 1) * t) 
    else:
        return formatted(avaliable[-1][-1] - 1)
    

def formatted(time):
    return f"{time // 60:02}:{time % 60:02}"

File: cjeongmin/셔틀버스.js
const START = 9 * 60;

const toMinutes = (time) => {
    const [hh, mm] = time.split(":").map(Number);
    return hh * 60 + mm;
};

const toHHMM = (minutes) => {
    const hh = String(Math.floor(minutes / 60)).padStart(2, "0");
    const mm = String(minutes % 60).padStart(2, "0");
    return `${hh}:${mm}`;
};

const canBoard = (n, t, m, x, crewTimes) => {
    let idx = 0;
    for (let i = 0; i < n; i++) {
        let capacity = m;
        const busTime = START + i * t;
        while (capacity > 0 && idx < crewTimes.length && crewTimes[idx] <= x && crewTimes[idx] <= busTime) {
            capacity--;
            idx++;
        }

        if (capacity > 0 && x <= busTime) return true;
    }
    return false;
};

function solution(n, t, m, timetable) {
    const crewTimes = timetable.map(toMinutes).sort((a, b) => a - b);

    let low = 0;
    let high = 23 * 60 + 59;
    let answer = START;
    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (canBoard(n, t, m, mid, crewTimes)) {
            answer = mid;
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }

    return toHHMM(answer);
}
